# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [158]:
def najdi_prvocisla(n, prvocisla = [2]):
    print(f"argument list parametru prvocisla obsahuje {prvocisla}")
    i = prvocisla[-1]
    while len(prvocisla) < n:
        i += 1
        for prvocislo in prvocisla:
            if i % prvocislo == 0:
                break
        else:
            prvocisla.append(i)
    return prvocisla

In [55]:
print(najdi_prvocisla(2))
print(najdi_prvocisla(2))
print(najdi_prvocisla(10))
print(najdi_prvocisla(40))

argument list parametru prvocisla obsahuje [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
argument list parametru prvocisla obsahuje [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
argument list parametru prvocisla obsahuje [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
argument list parametru prvocisla obsahuje [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173]



# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [105]:
%%writefile pi_approximation.py
import math
def proste_iterace(a,n):
    x_i = a
    for _ in range(n):
        x = (a/x_i + x_i)/2
        x_i = x
    return x_i

# vypocte obsah 6*2^n uhelniku, odhad pi
def odhad_pi(n):
    # inicializace hodnot, zaciname s rovnoramennym trojuhelnikem s delkou strany 1
    # a je delka ramene a vzdy bude rovna 1, b je delka zakladny, v je vyska, S je obsah aktualniho mnohouhelniku
    a, b, v, S = 1, 1, None, 0

    v = proste_iterace(a**2 - (b/2)**2, 100)
    if n == 0:
        S = 6 * 2**n * b * v / 2
        return S

    condition_check = False
    for i in range(1,n):
        b = proste_iterace((b/2)**2 + (1 - v)**2, 100)
        v = proste_iterace(a**2 - (b/2)**2, 100)
        S = 6 * 2**i * b * v / 2
        if (abs(S - math.pi) <= 10**-6) and (condition_check is not True):
            condition_check = True
            # print(f"Potrebujeme alespon {6 * 2**i} uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace {i}")
            continue


    return S

def odhad_pi_newton(n):
    a = (1 / 2) * (1 / (2 ** 3))
    suma = ((1 / ((2 * 1) + 1)) * a)
    for i in range(2, n + 1):
        a = a * ((2 * i - 3) / (2 * i)) * (1 / (2 ** 2))
        suma += ((1 / ((2 * i) + 1)) * a)
    pi = 12 * (-(proste_iterace(3, 1000) / 8) + (1 / 2) - suma)
    return pi

Writing pi_approximation.py


In [106]:
from pi_approximation import odhad_pi, odhad_pi_newton, proste_iterace

In [94]:
%time proste_iterace(123456123, 1000000)

CPU times: total: 31.2 ms
Wall time: 161 ms


11111.081090515

In [95]:
%%timeit
proste_iterace(123456123, 1000)

131 µs ± 45.7 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [96]:
%timeit odhad_pi_newton(20)

122 µs ± 37.6 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [97]:
%%timeit
odhad_pi_newton(20)

244 µs ± 78.7 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [98]:
%%time
odhad_pi_newton(20)

CPU times: total: 0 ns
Wall time: 520 µs


3.1415926535897936


# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [107]:
for n in range(20,36):
    pi = odhad_pi(n)
    print(f"{n=}, {pi=}")

Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=20, pi=3.1415926535877032
Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=21, pi=3.14159265358927
Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=22, pi=3.1415926535896617
Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=23, pi=3.14159265358976
Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=24, pi=3.1415926535897842
Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=25, pi=3.141592653589791
Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=26, pi=3.141592653589792
Potrebujeme alespon 6144 uhelnik, aby aproximace Pi byla s chybou mensi nez 10 ^ -6. Iterace 10
n=27, pi=3.1415926535897922
Potrebujeme al

In [108]:
for n in range(1,10):
    pi = odhad_pi_newton(n)
    print(f"{n=}, {pi=}")

n=1, pi=3.1519237886466844
n=2, pi=3.1425487886466845
n=3, pi=3.1417117350752557
n=4, pi=3.1416100098148387
n=5, pi=3.1415954446070975
n=6, pi=3.1415931337808694
n=7, pi=3.1415927403902137
n=8, pi=3.1415926698836163
n=9, pi=3.1415926567409396



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [114]:
def fib(n):
    prev, next = 0,1
    for i in range(0,n):
        yield prev
        prev, next = next, prev+ next

In [120]:
fibonacci_seq = fib(9)
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))
print(next(fibonacci_seq))

0
1
1
2
3
5
8
13
21


StopIteration: 

In [121]:
fibonacci_seq = fib(5)
for i in fibonacci_seq:
    print(i)

0
1
1
2
3


# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [123]:
num_list = [x for x in range(1,11)]
num_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [125]:
squared_list = [x**2 for x in range(1,11)]
squared_list

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [162]:
lst = [i for i in range(1,10) if ]

argument list parametru prvocisla obsahuje [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
True


# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [136]:
%%writefile test.txt
This is a test file

Writing test.txt


In [138]:
with open("test.txt", "r") as file:
    print(file.read())

This is a test file



In [142]:
with open("write.txt", "w") as file:
    file.write("Hello,\nworld\n")
    for i in range(1,11):
        file.write(f"{i: >6} {i**2: >6} {i**3: >6}\n")

# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [67]:
# kostra k doplnění, místa označená TODO doplňte

class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass
    
    def __add__(self, other):
        return Addition(self,other)
    
    def __sub__(self, other):
        # TODO (použijte Addition a Multiplication se zápornou konstantu)
        return Addition(self,Multiplication(-1, other))
    
    def __mul__(self, other):
        # TODO (použijte Multiplication)
        return Multiplication(self,other)
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
        self.argument = argument
    
    def __str__(self):
        return type(self).__name__ + "(" + str(self.argument) + ")"
        
class Addition(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
        return f"{self.levy_argument} + {self.pravy_argument}"
    def diff(self):
        # TODO
        return Addition(self.levy_argument.diff(), self.pravy_argument.diff())

    # def simplify(self):
    #     if self.pravy_argument.simplify() is None:
    #         return f"{self.levy_argument}"
    #     elif self.levy_argument.simplify() is None:
    #         return f"{self.pravy_argument}"
    
class Multiplication(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)*(pravy_argument)
        return f"({self.levy_argument})*({self.pravy_argument})"
    def diff(self):
        # TODO
        return Addition(Multiplication(self.levy_argument.diff(), self.pravy_argument), Multiplication(self.levy_argument, self.pravy_argument.diff()))
    
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        # TODO vypíšeme něco jako value
        return f"{self.value}"
    def diff(self):
        # TODO
        return "0"
    #
    # def simplify(self):
    #     if self.value == 0:
    #         return None
    
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        # TODO vypíšeme něco jako x   
        return "x"
    def diff(self):
        # TODO
        return "1"
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        # TODO vypíšeme něco jako argument^exponent
        return f"({self.argument})^{self.exponent}"
    def diff(self):
        # TODO
        return Multiplication(Multiplication(self.exponent, mocnina(self.argument, self.exponent - 1)), self.argument.diff())
    
class sin(UnaryOperator):
    def diff(self):
        # TODO
        return Multiplication(cos(self.argument), self.argument.diff())

class cos(UnaryOperator):
    def diff(self):
        # TODO
        return Multiplication(Multiplication(-1, sin(self.argument)), self.argument.diff())

class exp(UnaryOperator):
    def diff(self):
        # TODO
        return Multiplication(exp(self.argument), self.argument.diff())
    
class ln(UnaryOperator):
    def diff(self):
        # TODO
        return Multiplication(frc(self.argument), self.argument.diff())

class frc(UnaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako 1/(argument)
        return f"1/({self.argument})"
    
    def diff(self):
        # TODO
        return mocnina(self.argument, -1).diff()

    # def simplify(self):
    #     if self.argument < 0:

In [35]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

sin(x + cos((x)*(x))) + 1
(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [36]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


(x)^2 + (2)*(x) + 1 + (2)*((x)^2)
((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [37]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


(x)^2 + (2)*(x) + exp(x)
((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


In [39]:
x = identity()
f4 = frc(x)
print(f4)
print(f4.diff())

1/(x)
((-1)*((x)^-2))*(1)


In [68]:
f = x + cst(0)
print(f)
print(f.simplify())

x + 0
None


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...